utils/src/activityLog/Activity.ts

```diff
@@ -0,0 +1,87 @@
+*--------------------------------------------------------------------------------------------*/
+
+import { randomUUID } from "crypto";
+import { EventEmitter, TreeItemCollapsibleState } from "vscode";
+import * as types from '../../index';
+import { parseError } from "../parseError";
+import { AzExtTreeItem } from "../tree/AzExtTreeItem"
+
+export interface ActivityTreeItemOptions {
+    label: string;
+    contextValuePostfix?: string;
+    collapsibleState?: TreeItemCollapsibleState;
+    children?: (parent: types.AzExtParentTreeItem) => AzExtTreeItem[];
+}
+
+export abstract class ActivityBase implements types.Activity {
+
+    public readonly onStart: typeof this._onStartEmitter.event;
+    public readonly onProgress: typeof this._onProgressEmitter.event;
+    public readonly onSuccess: typeof this._onSuccessEmitter.event;
+    public readonly onError: typeof this._onErrorEmitter.event;
+
+    private readonly _onStartEmitter: EventEmitter<types.OnStartActivityData>;
+    private readonly _onProgressEmitter: EventEmitter<types.OnProgressActivityData>;
+    private readonly _onSuccessEmitter: EventEmitter<types.OnSuccessActivityData>;
+    private readonly _onErrorEmitter: EventEmitter<types.OnErrorActivityData>;
+
+    public running: boolean;
+    public done: boolean;
+    public error?: types.IParsedError;
+    public readonly task: types.ActivityTask;
+    public startedAtMs: number;
+    public readonly id: string;
+
+    public progress: { message?: string, increment?: number }[];
+
+    abstract initialState(): ActivityTreeItemOptions;
+    abstract successState(): ActivityTreeItemOptions;
+    abstract errorState(error: types.IParsedError): ActivityTreeItemOptions;
+
+    public constructor(task: types.ActivityTask) {
+        this.id = randomUUID();
+        this.done = false;
+        this.startedAtMs = Date.now();
+        this.task = task;
+        this.progress = [];
+
+        this._onStartEmitter = new EventEmitter();
+        this._onProgressEmitter = new EventEmitter();
+        this._onSuccessEmitter = new EventEmitter();
+        this._onErrorEmitter = new EventEmitter();
+
+        this.onStart = this._onStartEmitter.event;
+        this.onProgress = this._onProgressEmitter.event;
+        this.onSuccess = this._onSuccessEmitter.event;
+        this.onError = this._onErrorEmitter.event;
+    }
+
+    public report(progress: { message?: string; increment?: number }): void {
+        this.progress.unshift(progress);
+        this._onProgressEmitter.fire({ ...this.getState(), message: progress.message });
+    }
+
+    public async run(): Promise<void> {
+        try {
+            this._onStartEmitter.fire(this.getState());
+            await this.task({ report: (progress) => this.report(progress) });
+            this.done = true;
+            this._onSuccessEmitter.fire(this.getState());
+        } catch (e) {
+            this.error = parseError(e);
+            this.done = true;
```

Comment:
Small thing, but you can put the `this.done` in a finally clause.

Reply:
The finally clause isn't guaranteed to be executed since we re-throw the error in the catch block. To guarantee the finally block is always executed regardless of if an error is thrown in the catch block, you must return inside it. I just learned this, and I plan on us forgetting this behavior in the future, so I'm gonna leave it as-is.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/try...catch#returning_from_a_finally-block

Finally block will not execute:
```ts
try {
  	throw new Error('hi');
} catch (error) {
  	console.error(error);
  	throw error;
} finally {
  console.log('Finally block executed');
}
```

If we add a `return` to the finally block, it will always execute:
```ts
try {
  	throw new Error('hi');
} catch (error) {
  	console.error(error);
  	throw error;
} finally {
  	console.log('Finally block executed');
	return;
}
```

Reply:
Maybe a comment to reduce the chance of us forgetting this? :smile: